### ISSUE TITLE
Cache intermediate results in custom op for backwards pass

### ISSUE THREAD

**Issue by marsupial-tail** [tensorflow/tensorflow#25703]
Hello,

I wonder if there is a feature in Tensorflow which allows caching of intermediate results in a custom operation for the backwards computation, similar to the the ctx->save_for_backward interface in Pytorch. Does the C++ context object that we get to work with provide such a functionality?

Best,

Ziheng Wang

**Comment by jvishnuvardhan** [comment]
@marsupial-tail Could you describe more details about context of your problem. Thanks!

**Comment by marsupialtail** [comment]
Say I want to implement a custom tensorflow operation that fuses convolution and batch norm, which I believe from reading source code is not supported rn. I would like to cache the input to the batchnorm so I can compute gradients for this op on the back pass.

**Comment by marsupial-tail** [comment]
For example, I quote the cudnn documentation for the api cudnnBatchNormalizationForwardTraining: " resultSaveMean, resultSaveInvVariance
Outputs. Optional cache to save intermediate results computed during the forward pass. These buffers can be used to speed up the backward pass when supplied to the cudnnBatchNormalizationBackward() function. The intermediate results stored in resultSaveMean and resultSaveInvVariance buffers should not be used directly by the user. Depending on the batch normalization mode, the results stored in resultSaveInvVariance may vary. For the cache to work correctly, the input layer data must remain unchanged until the backward function is called. Note that both parameters can be NULL but only at the same time. In such a case intermediate statistics will not be saved, and cudnnBatchNormalizationBackward() will have to re-compute them. It is recommended to use this cache as the memory overhead is relatively small because these tensors have a much lower product of dimensions than the data tensors. "

If I were to indeed cache those in my op, how would I actually save these data pointers in Tensorflow?

**Comment by marsupial-tail** [comment]
Any updates?

**Comment by rmothukuru** [comment]
@marsupial-tail,
Sorry for the delayed response. Can you please refer the detailed documentation regarding [Custom Op](https://www.tensorflow.org/guide/create_op) and let us know if it helps? Thanks!

**Comment by qingchanghan** [comment]
Hi, I have the same problem.
In calculation of backward pass, how I can use the tensor allocated for forward pass by `allocate_persistent` function?
